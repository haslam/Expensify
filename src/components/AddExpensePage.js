import React from 'react';
import { connect } from "react-redux";
import ExpenseForm from './ExpenseForm';
import { addExpense } from "../actions/Expenses";

//always use an unconnected component for testing. Hence I'll add an export here
export class AddExpensePage extends React.Component {
  constructor(props) {
    super(props);
    this.onSubmit = this.onSubmit.bind(this);
  }
  onSubmit (expense) {
    //props.dispatch(addExpense(expense)) -- now, we can call onSubmit of mapped to props
    this.props.addExpense(expense);
    this.props.history.push('/');
  }
  render()  {
    return (
      <div>
        <h1>Add Expense</h1>
        <ExpenseForm
          onSubmit={this.onSubmit}
        />
      </div>
    )
  }
}

//For a more testable component, let's use mapDispatchToProps of connect. works similar as mapStateToProps but for Dispatch

const mapDispatchToProps = (dispatch) => ({
  addExpense: (expense) => dispatch(addExpense(expense))
})

//set first argument [state] to undefined.
export default connect(undefined, mapDispatchToProps)(AddExpensePage)