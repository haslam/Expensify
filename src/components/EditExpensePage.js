import React from 'react';
import { connect } from 'react-redux';
import { matchPath } from 'react-router-dom';
import ExpenseForm from './ExpenseForm';
import { startEditExpense, startRemoveExpense } from '../actions/Expenses';

export class EditExpensePage extends React.Component {
  constructor(props) {
    super(props);
    this.onSubmit = this.onSubmit.bind(this);
    this.onRemove = this.onRemove.bind(this);
  }
  onSubmit (expense) {
    this.props.startEditExpense(this.props.expense.id, expense);
    this.props.history.push('/');
    console.log('Updated', expense) 
  }
  onRemove () {
    this.props.startRemoveExpense({ id: this.props.expense.id} );
    console.log('Removed', this.props.expense.id);
    this.props.history.push('/');
  }
  render() {
  console.log('edit props ', this.props);
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h1>Edit Expense</h1>
            <p>Easily make changes to the selected expense.</p> 
          </div>
        </div>
        <div className="content-container">
        <ExpenseForm
          expense={this.props.expense}
          onSubmit={this.onSubmit}
        />
        <button className="button button--secondary" onClick={this.onRemove}>Remove Expense</button>
        </div> 
      </div>
    )
  }
}

const mapDispatchToProps = (dispatch, props) => ({
  startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
  startRemoveExpense: (data) => dispatch(startRemoveExpense(data))
})

const mapStateToProps = (state, props) => ({
  expense: state.expenses.find((expense) => {
  const match = matchPath(props.location.pathname, {
    path: '/edit/:id',
    exact: true,
    strict: false
  })
  return expense.id === match.params.id
  })
})
export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage)

