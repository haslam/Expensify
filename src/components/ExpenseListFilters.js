import React from 'react';
import moment from 'moment';
import { connect } from "react-redux";
import { DateRangePicker } from "react-dates";
import { setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate } from '../actions/Filters'



export class ExpenseListFilters extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      calenderFocused: null,
      filterText: '',
      filterSortBy: '',
      startDate: moment().startOf('month'),
      endDate: moment().endOf('month'),
    };
    this.onDatesChange = this.onDatesChange.bind(this);
    this.onFocusChange = this.onFocusChange.bind(this);
    this.onTextChange = this.onTextChange.bind(this);
    this.onSortChange = this.onSortChange.bind(this);
  }
  onTextChange (e) {
    const sT = e.target.value;
    this.setState(() => ({ filterText: sT}))
    this.props.setTextFilter(sT);
  }
  onSortChange(e) {
    const sortVal = e.target.value;
    this.setState(() => ({ filterSortBy: sortVal}))
    if (sortVal === 'date') {
      this.props.sortByDate();
    }
    else if (sortVal === 'amount') {
      this.props.sortByAmount();
    }
  }
  onDatesChange ({ startDate, endDate }) {
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  }
  onFocusChange (calenderFocused) {
    this.setState(() => ({ calenderFocused }));
  }

  render() {
   return (
  <div className="content-container">
  <div className="input-group">
    <div className="input-group__item">
      <input type="text" 
        value={this.state.filterText}
        onChange={this.onTextChange}
        className="text-input"
        placeholder="Search expenses"
      />
    </div>
    <div className="input-group__item">
      <select 
        value={this.state.filterSortBy} 
        onChange={this.onSortChange}
        className="select"
      >
        <option value="date">Date</option>
        <option value="amount">Amount</option>
      </select>
    </div>
    <div className="input-group__item">
      <DateRangePicker
        startDate={this.props.filters.startDate} // momentPropTypes.momentObj or null,
        //startDateId={Math.floor((1 + Math.random() * 0x10000)).toString(16).substring(1)} // PropTypes.string.isRequired,
        startDateId={String(this.props.filters.startDate)}
        endDate={this.props.filters.endDate} // momentPropTypes.momentObj or null,
        //endDateId={Math.floor((1 + Math.random() * 0x10000)).toString(16).substring(1)} // PropTypes.string.isRequired,
        endDateId={String(this.props.filters.endDate)}
        onDatesChange={this.onDatesChange}
        //   this.props.setStartDate(startDate);
        //   this.props.setEndDate(endDate);
        //   } // PropTypes.func.isRequired,
        // } 
        focusedInput={this.state.calenderFocused} // PropTypes.oneOf([START_DATE, END_DATE]) or null,
        onFocusChange={calenderFocused => this.setState({ calenderFocused })} // PropTypes.func.isRequired,
        numberOfMonths={1}
        showDefaultInputIcon={true}
        showClearDates={true}
      />
    </div>
  </div>
  </div>
    );
  }
}

const mapStateToProps = (state) => ({
    filters: state.filters
})

const mapDispatchToProps = (dispatch, props) => ({
  setStartDate: (startDate) => dispatch(setStartDate(startDate)),
  setEndDate: (endDate) => dispatch(setEndDate(endDate)),
  setTextFilter: (text) => dispatch(setTextFilter(text)),
  sortByAmount: () => dispatch(sortByAmount()),
  sortByDate: () => dispatch(sortByDate())
})
export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);

{/* <DateRangePicker 
startDate={this.props.filters.startDate}
endDate={this.props.filters.endDate}
focusedInput={this.state.calenderFocused}
onFocusChange={this.onFocusChange}
showClearDates={true}
numberOfMonths={1}
isOutsideRange={() => false}      
/> */}