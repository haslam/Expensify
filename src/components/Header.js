import React from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import { beginLogout } from '../actions/auth'
import hashEmail from '../helper/hashing'

export const Header = ({ beginLogout, email, displayName }) => (
  <header className="header">
    <div className="content-container">
      <div className="header__content">
        <div className="show-for-mobile">
          <button className="mobile__hamburger" role="menu" title="Navigation menu hamburger" onClick={openNavigation}>{`\u{1F354}`}</button>
        </div>
        <Link className="header__title" to="/dashboard">
          <h1>Expensified</h1>
        </Link>
        
        <div className="header__profile show-for-desktop">
          <div className="header__avatar">
            <img className="avatar avatar--sphere" src={`https://gravatar.com/avatar/${hashEmail(email)}?s=200`}/>
            <small style={{paddingRight: "1rem", color: "#dcdcdc"}}>{displayName}</small>
          </div>
          <div style={{ alignSelf: "center" }}>
            <button className="button button--trans" onClick={beginLogout}>Logout</button>
          </div>  
        </div>
      </div>
    </div>
  </header>
)

const openNavigation = () => {
  document.body.classList.toggle('mobile__nav--on');
}

const mapDispatchToProps = (dispatch) => ({
  beginLogout: () => dispatch(beginLogout())
})

const mapStateToProps = (state) => ({
  email: state.auth.email,
  displayName: state.auth.displayName,
})
export default connect(mapStateToProps, mapDispatchToProps)(Header)

{/* <NavLink to="/create" activeClassName="is-active">Create Expense</NavLink> */}
