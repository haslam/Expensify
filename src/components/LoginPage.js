import React from "react";
import { connect } from "react-redux";
import { beginLogin } from "../actions/auth";
export const LoginPage = ({ beginLogin }) => (
  <div>
    <h2>Login with your username and password</h2>
      <button onClick={ beginLogin }>Login</button>
  </div>
)


const mapDispatchToProps = (dispatch) => ({
  beginLogin: () => dispatch(beginLogin())
})

export default connect(undefined, mapDispatchToProps)(LoginPage)

{/* <input type="text" name="username" placeholder="username"/>
<input type="password" name="password" placeholder="******" /> */}