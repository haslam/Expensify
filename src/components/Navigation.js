import React from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faTachometerAlt, faMoneyCheckAlt, faPoll, faChartLine } from "@fortawesome/free-solid-svg-icons";
import { beginLogout } from '../actions/auth'
import hashEmail from '../helper/hashing'

export const Navigation = ({ beginLogout, email, displayName}) => {

  return (
    <div className="mobile-navigation mobile-navigation--left">
      <div className="navigation__profile">
        <img className="avatar avatar--sphere" src={`https://gravatar.com/avatar/${hashEmail(email)}?s=200`}/>
        <small>{displayName}</small>
        <button className="button__logout--left" onClick={beginLogout}>Logout</button>
      </div>
      <div className="navigation__tabs">
        <ul role="navigation" className="tabs-menu">
          <li className="tabs-menu__list">
            <Link to="/dashboard" activeClassName="tm-active">
              <FontAwesomeIcon icon={faTachometerAlt} size="2x" />
            </Link>
          </li>
          <li className="tabs-menu__list">
            <Link to="/expenses" activeClassName="tm-active">
              <FontAwesomeIcon icon={faMoneyCheckAlt} size="2x" />
            </Link>
          </li>
          <li className="tabs-menu__list">
            <Link to="/report" activeClassName="tm-active">
              <FontAwesomeIcon icon={faPoll} size="2x" />
            </Link>
          </li>
        </ul>
      </div>
    </div>
  )
}

const mapDispatchToProps = (dispatch) => ({
  beginLogout: () => dispatch(beginLogout())
})
const mapStateToProps = (state) => ({
  email: state.auth.email,
  displayName: state.auth.displayName
})

export default connect(mapStateToProps, mapDispatchToProps)(Navigation)