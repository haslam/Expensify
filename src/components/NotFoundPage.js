import React from 'react';

export default () => (
  <p>
    You've hit a wrong path
  </p>
)