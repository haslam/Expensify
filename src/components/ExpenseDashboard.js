import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';
import ExpensesSummary from "./ExpensesSummaries";
import AddExpense from './AddExpensePage'
import Header from './Header';

const ExpenseDashboardPage = () => (
  <div>
    <Header />
    <ExpensesSummary />
    <ExpenseListFilters />
    <ExpenseList />
  </div>
)

export default ExpenseDashboardPage
