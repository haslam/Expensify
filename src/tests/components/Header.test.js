import React from "react";
import { shallow } from "enzyme";
import { Header } from "../../components/Header";
import hashEmail from '../../helper/hashing';

const email = 'test@example.com';

// beforeEach(() => {
//   logout = jest.fn();
//   wrapper = shallow(<Header beginLogout={() => {}}/>);
// })
test('should render Header correctly', () => {
const wrapper = shallow(<Header email={email}beginLogout={() => {}}/>);

  //use enzymn-to-json library to enhance enzyme with snapshot utility
  expect(wrapper).toMatchSnapshot();
 // expect(wrapper.find('h1').text()).toBe('Expensify');

  // const renderer = new ReactShallowRenderer();
  // renderer.render(<Header />);
  // expect(renderer.getRenderOutput()).toMatchSnapshot();
})

test('should call beginLogout on button click', () => {
  const beginLogout = jest.fn(); //spy
  const wrapper = shallow(<Header email={email} beginLogout={beginLogout} />);
  wrapper.find('.button--trans').simulate('click');
  expect(beginLogout).toHaveBeenCalled()
})

