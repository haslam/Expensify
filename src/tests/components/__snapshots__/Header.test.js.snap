// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Header correctly 1`] = `
<header
  className="header"
>
  <div
    className="content-container"
  >
    <div
      className="header__content"
    >
      <div
        className="show-for-mobile"
      >
        <button
          className="mobile__hamburger"
          onClick={[Function]}
          role="menu"
          title="Navigation menu hamburger"
        >
          üçî
        </button>
      </div>
      <Link
        className="header__title"
        replace={false}
        to="/dashboard"
      >
        <h1>
          Expensified
        </h1>
      </Link>
      <div
        className="header__profile show-for-desktop"
      >
        <div
          className="header__avatar"
        >
          <img
            className="avatar avatar--sphere"
            src="https://gravatar.com/avatar/55502f40dc8b7c769880b10874abc9d0?s=200"
          />
          <small
            style={
              Object {
                "color": "#dcdcdc",
                "paddingRight": "1rem",
              }
            }
          />
        </div>
        <div
          style={
            Object {
              "alignSelf": "center",
            }
          }
        >
          <button
            className="button button--trans"
            onClick={[Function]}
          >
            Logout
          </button>
        </div>
      </div>
    </div>
  </div>
</header>
`;
